source/common/http/header_map_impl.h

```diff
@@ -180,59 +188,89 @@ class HeaderMapImpl : NonCopyable {
+      if (!lazy_map_.empty()) {
+        lazy_map_[i->key().getStringView()].push_back(i);
+      }
       if (!is_pseudo_header && pseudo_headers_end_ == headers_.end()) {
         pseudo_headers_end_ = i;
       }
       return i;
     }
 
-    std::list<HeaderEntryImpl>::iterator erase(std::list<HeaderEntryImpl>::iterator i) {
+    HeaderNode erase(HeaderNode i, bool remove_from_map) {
       if (pseudo_headers_end_ == i) {
         pseudo_headers_end_++;
       }
+      if (remove_from_map) {
+        lazy_map_.erase(i->key().getStringView());
+      }
       return headers_.erase(i);
     }
 
     template <class UnaryPredicate> void remove_if(UnaryPredicate p) {
+      // Lazy map isn't used, remove items from the list directly
       headers_.remove_if([&](const HeaderEntryImpl& entry) {
         const bool to_remove = p(entry);
         if (to_remove) {
           if (pseudo_headers_end_ == entry.entry_) {
             pseudo_headers_end_++;
           }
+          if (!lazy_map_.empty()) {
+            auto& values_vec = lazy_map_[entry.key().getStringView()];
+            if (values_vec.size() == 1) {
+              lazy_map_.erase(entry.key().getStringView());
+            } else {
+              values_vec.erase(std::remove_if(values_vec.begin(), values_vec.end(),
+                                              [&](HeaderNode it) { return it == entry.entry_; }),
+                               values_vec.end());
+            }
+          }
         }
         return to_remove;
       });
     }
 
+    // Creates and populates a map if the number of headers is at least HEADER_MAP_SIZE_THRESHOLD.
+    // Returns true if a map was created
+    bool maybeMakeMap();
+
+    // Removes a given key and its values from the HeaderList, and returns the number of bytes
+    // that were removed
+    size_t remove(absl::string_view key);
+
     std::list<HeaderEntryImpl>::iterator begin() { return headers_.begin(); }
     std::list<HeaderEntryImpl>::iterator end() { return headers_.end(); }
     std::list<HeaderEntryImpl>::const_iterator begin() const { return headers_.begin(); }
     std::list<HeaderEntryImpl>::const_iterator end() const { return headers_.end(); }
     std::list<HeaderEntryImpl>::const_reverse_iterator rbegin() const { return headers_.rbegin(); }
     std::list<HeaderEntryImpl>::const_reverse_iterator rend() const { return headers_.rend(); }
+    HeaderLazyMap::iterator mapFind(absl::string_view key) { return lazy_map_.find(key); }
+    HeaderLazyMap::iterator mapEnd() { return lazy_map_.end(); }
     size_t size() const { return headers_.size(); }
     bool empty() const { return headers_.empty(); }
     void clear() {
       headers_.clear();
       pseudo_headers_end_ = headers_.end();
+      lazy_map_.clear();
     }
 
   private:
     std::list<HeaderEntryImpl> headers_;
-    std::list<HeaderEntryImpl>::iterator pseudo_headers_end_;
+    HeaderNode pseudo_headers_end_;
+    HeaderLazyMap lazy_map_;
```

Comment:
`#if HEADER_MAP_SIZE_THRESHOLD > 0` here?

Reply:
If it won't be declared here, we'll need to encompass all the rest of the code that accesses the map with a macro.
I prefer not to add it at the moment, to keep the code more readable, but let me know if you think otherwise.

Reply:
How long do we expect to leave HEADER_MAP_THRESHOLD at 0? If it will be immediately followed up by setting it to something higher in another PR then I agree it's definitely worth not optimizing for.

Reply:
I tend to agree.
I expect most of the header-maps (specifically request/response) to have at least 2/3 headers, and so far I think that the microbenchmarks/benchmarks suggest that in these cases the threshold should be larger than 0.

Reply:
I thought 0 was a special-case in your ifdef to say "don't use maps" but I could be mistaken.

But the question of thresholding is what's the cutover point where a map is a win, compared to linear walk over the vector. I thought it might be around 5-10, but it depends on which map and probably usage pattern.

Reply:
I removed the macro for the threshold, and it is now set as a runtime flag (which is set to uint32::max by default).
The code uses the flat_hash_map by default. If HEADERMAP_TYPE_MULTIMAP is set, uses std::mutlimap, and if both HEADERMAP_TYPE_MULTIMAP and MULTIMAP_ABSL_BTREE are set, uses the btree implementation.

It seems that flat_hash_map with threshold=3 gives a good performance on most benchmarks, and the benchmark that suffers the most is the RemoveInline (~1.16x slower).